csrc/runtime/fusion_kernel_runtime.cpp

```diff
@@ -361,71 +361,89 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
-          const std::lock_guard<std::mutex> lock(
-              thread_pool_error_message_mutex);
-          std::stringstream ss;
-          ss << thread_pool_error_message << "\nError from segmentation group "
-             << group_to_run->groupId() << ": " << e.what() << "\n";
-          thread_pool_error_message = ss.str();
-        }
-      });
-    }
+        c10::cuda::CUDAGuard dg(args.getDeviceIndex());
+        c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
+        compileKernel(group_runtime_inputs, group_to_run, hic.get());
+      } else {
+        hir::HostIrContainer* hic_ptr = hic.get();
+        // launch compileKernel thread here
+        getThreadPool()->run([this,
+                              args,
+                              group_runtime_inputs,
+                              group_to_run,
+                              &detect_exception_in_thread_pool,
+                              &thread_pool_error_message,
+                              &thread_pool_error_message_mutex,
+                              hic_ptr]() {
+          FUSER_PERF_SCOPE("FusionKernelRuntime::compileFusionParallel");
+          try {
+            c10::cuda::CUDAGuard dg(args.getDeviceIndex());
+            c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
+            compileKernel(group_runtime_inputs, group_to_run, hic_ptr);
+          } catch (const std::exception& e) {
+            // Set flag inside lambda so we can throw an exception after thread
+            // pool completes its work.
+            detect_exception_in_thread_pool.store(true);
+            const std::lock_guard<std::mutex> lock(
+                thread_pool_error_message_mutex);
+            std::stringstream ss;
+            ss << thread_pool_error_message
+               << "\nError from segmentation group " << group_to_run->groupId()
+               << ": " << e.what() << "\n";
+            thread_pool_error_message = ss.str();
+          }
+        });
+      }
 
-    auto fusion_to_run = segmented_fusion_->makeFusion(group_to_run).second;
-    auto group_runtime_outputs =
-        inferOutputSizes(fusion_to_run.get(), group_runtime_inputs);
+      auto fusion_to_run = segmented_fusion_->makeFusion(group_to_run).second;
+      auto group_runtime_outputs =
+          inferOutputSizes(fusion_to_run.get(), group_runtime_inputs);
 
-    // map output args to tensor map
-    args_manager.updateWithSegmentOutputs(
-        group_to_run->outputs(), group_runtime_outputs, run_order_id);
+      // map output args to tensor map
+      args_manager.updateWithSegmentOutputs(
+          group_to_run->outputs(), group_runtime_outputs, run_order_id);
+    }
+  } catch (const std::exception& e) {
+    // Before clearning up unique_ptr-backed resources such as
+    // SegmentedGroup, make sure all threads are done as they may
+    // be still using the resources.
+    getThreadPool()->waitWorkComplete();
```

Comment:
How is this different from the waitWorkComplete on line 451?

Reply:
The main thread might exit before reaching that line. That was what's happening with Christian's WIP PR. Specifically, he had some changes that caused an exception at `inferOuputSizes` at line 435.

Reply:
Got it! 

One simplification is to collect inputs and compile kernels in separate loops:
```
kernels_to_compile = []

for each segment in topological order:
  collect runtime inputs
  kernels_to_compile.append((segment ID, runtime inputs))
  infer output sizes
  update the Val-to-tensor map (aka args_manager)

parallel for each kernel in kernels_to_compile:
  compile the kernel using the recorded runtime inputs
```

IIUC, `group_runtime_inputs` are cheap, meta tensors, so keeping all of them alive won't be a problem.